### PR TITLE
Feat/data validation emails rebased

### LINKED ISSUE #199: data validation emails to boards
### User story: As a state administrator, I want to receive an email about my state's data upload so that I know whether the data from my state has uploaded successfully or if there are any errors. 
- email board contact addresses with info on validation errors
- check for errors every night
- If no errors, send that email once per week
- use existing email template

### PR DESCRIPTION
**Pull request by landonshumway-ia**
This PR is #349 rebased with the development branch to resolve some egregious merge conflicts. Keeping it on this separate branch to review with @jusdino.

Added an ingest event persistence feature to collect and report on license activity. Then added a nightly/weekly ingest failure / validation error report sender, which will deliver summary reports to the jurisdiction's configured operations contact.

In order to clearly distinguish between Python and NodeJS based lambdas, this PR introduces a file structure change to place all Python lambdas under a `lambda/python` directory, as well as the Node lambdas under a `lambdas/nodejs` directory.

### Requirements List
- New TypeScript lambda, with included package.json, managed by yarn

### Description List
- New TypeScript lambda which sends email notifications to jurisdictions for validation errors in data uploads
- Added CDK Reporting stack with constructs for reporting events.
- Lambda directory structure updated to clearly distinguish between Node and Python lambdas.

### Testing List
- `yarn test` should run without errors or warnings
- `yarn build` should run without errors or warnings
- Code review

Closes #199

### CODE REVIEW

**Comment by landonshumway-ia** [comment]
@jusdino if everything looks good here, we can either send this PR to Josh for review or move these changes into your current branch.

**Review by jusdino — APPROVED** [review]
Looks like you got everything in the rebase. I've just got to chase down the NodeJS build failure you mentioned in #349 before we send this along.

**Comment by jusdino** [comment]
Ok, @landonshumway-ia , ☝️ that was a non-trivial change to fix the `NodeJSFunction` issue, but I think it's for the better. I wouldn't mind one last sanity check from you before we pass this one along.

**Review by landonshumway-ia — COMMENTED** [review]
LGTM 👍

**Comment by landonshumway-ia** [comment]
I've confirmed the local deployment work for my sandbox. @jlkravitz This is ready for CSG review. Thanks

File: backend/compact-connect/lambdas/python/data-events/tests/function/__init__.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    def setUp(self):  # noqa: N801 invalid-name
+        super().setUp()
+
+        self.build_resources()
+
+        from cc_common import config
+
+        config.config = config._Config()  # noqa: SLF001 protected-access
+        self.config = config.config
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/data-events/tests/function/__init__.py:26]: isn't this already done by the super class?

File: backend/compact-connect/lambdas/python/provider-data-v1/tests/unit/test_handlers/test_bulk_upload.py [hunk 2]
```diff
@@ -10,15 +11,17 @@
 
 class TestProcessS3Event(TstLambdas):
     @patch('handlers.bulk_upload.process_bulk_upload_file', autospec=True)
-    @patch('handlers.bulk_upload.config', autospec=True)
+    # We can't autospec because it causes the patch to evaluate properties that look up environment variables that we
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/provider-data-v1/tests/unit/test_handlers/test_bulk_upload.py:14]: Helpful comment!

File: backend/compact-connect/pipeline/backend_stage.py [hunk 3]
```diff
@@ -63,3 +64,15 @@ def __init__(
             environment_name=environment_name,
             persistent_stack=self.persistent_stack,
         )
+
+        # Reporting depends on emails, which depend on having a domain name. If we don't configure a HostedZone
+        # we won't bother with this whole stack.
```
Comment (jlkravitz) [inline backend/compact-connect/pipeline/backend_stage.py:69]: Good context, helpful. Thank you!

File: backend/compact-connect/stacks/persistent_stack/data_event_table.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            batch_size=10,
+            encryption_key=encryption_key,
+            alarm_topic=alarm_topic,
+        )
+
+        Rule(
+            self,
+            'EventReceiverRule',
+            event_bus=event_bus,
+            # We will ignore ingest events, since they have sensitive data, but will store everything else
```
Comment (jlkravitz) [inline backend/compact-connect/stacks/persistent_stack/data_event_table.py:105]: Does this mean we won't get alerts on ingest event failures?

File: backend/compact-connect/stacks/reporting_stack.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    def __init__(self, scope: Construct, construct_id: str, *, persistent_stack: ps.PersistentStack, **kwargs):
+        super().__init__(scope, construct_id, **kwargs)
+        self._add_ingest_event_reporting_chain(persistent_stack)
+
+    def _add_ingest_event_reporting_chain(self, persistent_stack: ps.PersistentStack):
+        from_address = f'noreply@{persistent_stack.user_email_notifications.email_identity.email_identity_name}'
+
+        # We use a Node.js function in this case because the tool we identified for email report generation,
+        # EmailBuilderJS, is in Node.js. To make utilizing the tool as simple as possible, we opted to not mix
+        # languages in the Lambda.
```
Comment (jlkravitz) [inline backend/compact-connect/stacks/reporting_stack.py:28]: 🥳

File: backend/compact-connect/stacks/reporting_stack.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            self.format_arn(
+                partition=self.partition,
+                service='ses',
+                region=self.region,
+                account=self.account,
+                resource='configuration-set',
+                resource_name=persistent_stack.user_email_notifications.config_set.configuration_set_name,
+            ),
+        ]
+        # We'll assume that, if it is a sandbox environment, they're in the SES sandbox
```
Comment (jlkravitz) [inline backend/compact-connect/stacks/reporting_stack.py:58]: I know I've asked before, but remind me what SES is? Can we spell that out here?

**Review by jlkravitz — COMMENTED** [review]
A few comments/questions!

File: backend/compact-connect/README.md [hunk 7]
```diff
@@ -91,6 +98,16 @@ duplication and ensure consistency across the app.
 
 To execute the tests, simply run `bin/run_tests.sh` from the `backend` directory.
 
+## Documentation
+[Back to top](#compact-connect---backend-developer-documentation)
+
+Keeping documentation current is an important part of feature development in this project. If the feature involves a non-trivial amount of architecture or other technical design, be sure that the design and considerations are captured in the [design documentation](./docs/design). If any updates are made to the API, be sure to follow these steps to keep the documentation current:
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/README.md:104]: [It exists](https://github.com/InspiringApps/JCC-CompactConnect/tree/feat/data-validation-emails-rebased/backend/compact-connect/docs/design) and if you open the README in GitHub, you'll find that the link works.

[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/README.md:104]: There are couple of diagrams in the linked location, but neither is really a 'high level' architecture. They're more just diagramming design choices for particular components.

File: backend/compact-connect/lambdas/nodejs/.gitignore [hunk 8]
```diff
@@ -0,0 +1,6 @@
+lib/**/*.mjs
+lib/**/*.js
+lib/**/*.js.map
+bin/**/*.mjs
+bin/**/*.js
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/lambdas/nodejs/.gitignore:5]: Yep. This is TypeScript, so the `.js`, etc files are all just build artifacts.

File: backend/compact-connect/lambdas/nodejs/ingest-event-reporter/.gitignore [hunk 9]
```diff
@@ -0,0 +1,3 @@
+*.mjs
+*.js
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/lambdas/nodejs/ingest-event-reporter/.gitignore:2]: For this folder, yes.

File: backend/compact-connect/lambdas/nodejs/lib/event-client.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                (lastWeek.valueOf()/1000).toString()
+            ),
+            Number.parseInt(
+                (today.valueOf()/1000).toString()
+            )
+        ];
+    }
+
+    /*
+     * Queries the data event table for validation errors by looking for each comp
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/lambdas/nodejs/lib/event-client.ts:76]: Weird. Looks like the end of my comment got chopped off somehow. Will fix.

File: backend/compact-connect/lambdas/nodejs/tests/environment-variables-service.test.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+        environmentVariables = new EnvironmentVariablesService();
+    });
+
+    it('should produce a logger with debug log level', async () => {
+        const logger = new Logger({ logLevel: environmentVariables.getLogLevel() });
+
+        logger.debug('Test!');
+
+        // Verify the send method was called
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/lambdas/nodejs/tests/environment-variables-service.test.ts:21]: Stray (and wrong) comment. Removing. The debug line is not strictly necessary, but it was useful when I was looking at the tests to see that the log message actually shows up.

File: backend/compact-connect/lambdas/nodejs/tests/lambda.test.ts [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+            dynamoDBClient: asDynamoDBClient(mockDynamoDBClient),
+            sesClient: asSESClient(mockSESClient)
+        });
+
+        const resp = await lambda.handler(
+            SAMPLE_NIGHTLY_EVENT,
+            SAMPLE_CONTEXT
+        );
+
+        // Verify the DynamoDB client was called correctly
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/lambdas/nodejs/tests/lambda.test.ts:131]: Yes. You're suggesting we should also check for that call here? I'll add that in.

File: backend/compact-connect/lambdas/nodejs/tests/report-emailer.test.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            SAMPLE_SORTABLE_VALIDATION_ERROR_RECORDS
+        );
+
+        const flattenedErrors: string[] = sorted.flatMap((record) => record.errors.dateOfRenewal);
+
+        expect(flattenedErrors).toEqual([
+            'Row 4, 5:47',
+            'Row 5, 4:47',
+            'Row 5, 5:47'
+        ]);
```
[the remark this answers is not in the document]
Comment (jusdino) [reply backend/compact-connect/lambdas/nodejs/tests/report-emailer.test.ts:124]: Yes, this test is only testing that the `sortValidationErrors` method is sorting in the way we expect, not testing anything about the full validation report generation.

File: backend/compact-connect/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jlkravitz) [reply backend/compact-connect/README.md:104]: oops- I think I was looking from the root directory, not from `backend/compact-connect`.

File: backend/compact-connect/lambdas/nodejs/tests/lambda.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jlkravitz) [reply backend/compact-connect/lambdas/nodejs/tests/lambda.test.ts:131]: Great! Mostly was just wondering if there was a substantive reason that that wasn't checked here- but sounds like not. I think it makes sense to check in both places.

File: backend/compact-connect/lambdas/nodejs/tests/report-emailer.test.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (jlkravitz) [reply backend/compact-connect/lambdas/nodejs/tests/report-emailer.test.ts:124]: Ah got it! Thanks for the clarification.

File: backend/compact-connect/lambdas/python/data-events/tests/function/__init__.py (hunk 1 above)
Comment (jusdino) [reply backend/compact-connect/lambdas/python/data-events/tests/function/__init__.py:26]: Good catch. Removing.

File: backend/compact-connect/stacks/persistent_stack/data_event_table.py (hunk 4 above)
Comment (jusdino) [reply backend/compact-connect/stacks/persistent_stack/data_event_table.py:105]: No, it just means that we're not storing them in the events database for this feature (which will change soon, anyway).

**Comment by jusdino** [comment]
@jlkravitz , I think I've addressed each of your comments, now 🙂

File: backend/compact-connect/stacks/persistent_stack/data_event_table.py (hunk 4 above)
Comment (jlkravitz) [reply backend/compact-connect/stacks/persistent_stack/data_event_table.py:105]: Ah- got it.

**Review by jlkravitz — APPROVED** [review]
@jusdino Looks great!

@isabeleliassen This is good to merge.

**Review by isabeleliassen — APPROVED** [review]

**Pull request merged**